File: src/python/analysis.py
#!/usr/bin/env python3
"""
Northwestern MLDS 400 Assignment 3
Titanic Survival Analysis - Python Implementation
"""

import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import os

def main():
    print("TITANIC SURVIVAL ANALYSIS")
    print("Northwestern MLDS 400 Assignment 3")
    print("=" * 40)
    
    # Q13: Load training data
    print("\n[Q13] Loading training data")
    try:
        train = pd.read_csv("src/data/train.csv")
        print(f"Data shape: {train.shape}")
        print("First 5 rows:")
        print(train.head())
        print(f"Survival rate: {train['Survived'].mean():.3f}")
    except FileNotFoundError:
        print("Error: train.csv not found in src/data/")
        return
    
    # Q14: Feature engineering
    print("\n[Q14] Feature engineering")
    
    train['FamilySize'] = train['SibSp'] + train['Parch'] + 1
    print("Created FamilySize feature")
    
    train['IsAlone'] = (train['FamilySize'] == 1).astype(int)
    print("Created IsAlone feature")
    
    train['Sex_num'] = (train['Sex'] == 'male').astype(int)
    print("Applied gender encoding")
    
    train['Age'].fillna(train['Age'].median(), inplace=True)
    train['Embarked'].fillna(train['Embarked'].mode()[0], inplace=True)
    print("Filled missing values")
    
    # Q15: Train model
    print("\n[Q15] Model training")
    
    features = ['Pclass', 'Sex_num', 'Age', 'Fare', 'FamilySize', 'IsAlone']
    X = train[features]
    y = train['Survived']
    
    model = LogisticRegression(random_state=42, max_iter=1000)
    model.fit(X, y)
    print("Logistic regression trained")
    
    # Q16: Calculate accuracy
    print("\n[Q16] Training accuracy")
    
    pred = model.predict(X)
    acc = accuracy_score(y, pred)
    print(f"Accuracy: {acc:.4f}")
    
    # Q17: Process test data
    print("\n[Q17] Test data processing")
    
    try:
        test = pd.read_csv("src/data/test.csv")
        print(f"Test shape: {test.shape}")
        
        test['FamilySize'] = test['SibSp'] + test['Parch'] + 1
        test['IsAlone'] = (test['FamilySize'] == 1).astype(int)
        test['Sex_num'] = (test['Sex'] == 'male').astype(int)
        
        test['Age'].fillna(train['Age'].median(), inplace=True)
        test['Fare'].fillna(train['Fare'].median(), inplace=True)
        test['Embarked'].fillna(train['Embarked'].mode()[0], inplace=True)
        
        print("Applied transformations")
        
    except FileNotFoundError:
        print("Error: test.csv not found")
        return
    
    # Q18: Generate predictions
    print("\n[Q18] Predictions generated")
    
    X_test = test[features]
    predictions = model.predict(X_test)
    
    submission = pd.DataFrame({
        'PassengerId': test['PassengerId'],
        'Survived': predictions
    })
    
    os.makedirs("outputs", exist_ok=True)
    submission.to_csv("outputs/python_predictions.csv", index=False)
    
    print(f"Predictions for {len(predictions)} passengers")
    print(f"Survival rate: {predictions.mean():.3f}")
    print("Saved to outputs/python_predictions.csv")
    
    print("\nPython analysis completed")

if __name__ == "__main__":
    main()

File: src/r/analysis.R
#!/usr/bin/env Rscript
# Northwestern MLDS 400 Assignment 3
# Titanic Survival Analysis - R Implementation

library(readr)
library(dplyr)

cat("TITANIC SURVIVAL ANALYSIS\n")
cat("Northwestern MLDS 400 Assignment 3\n")
cat(rep("=", 40), "\n")

# Q13: Load training data
cat("\n[Q13] Loading training data\n")

if (!file.exists("src/data/train.csv")) {
    cat("Error: train.csv not found in src/data/\n")
    quit(status = 1)
}

train <- read_csv("src/data/train.csv", show_col_types = FALSE)
cat("Data shape:", paste(dim(train), collapse = " x "), "\n")
cat("First 5 rows:\n")
print(head(train, 5))
cat("Survival rate:", sprintf("%.3f", mean(train$Survived)), "\n")

# Q14: Feature engineering
cat("\n[Q14] Feature engineering\n")

train$FamilySize <- train$SibSp + train$Parch + 1
cat("Created FamilySize feature\n")

train$IsAlone <- as.numeric(train$FamilySize == 1)
cat("Created IsAlone feature\n")

train$Sex_num <- as.numeric(train$Sex == "male")
cat("Applied gender encoding\n")

train$Age[is.na(train$Age)] <- median(train$Age, na.rm = TRUE)
train$Embarked[is.na(train$Embarked)] <- names(sort(table(train$Embarked), decreasing = TRUE))[1]
cat("Filled missing values\n")

# Q15: Train model
cat("\n[Q15] Model training\n")

features <- c("Pclass", "Sex_num", "Age", "Fare", "FamilySize", "IsAlone")
model_data <- train[c(features, "Survived")]

model <- glm(Survived ~ ., data = model_data, family = binomial)
cat("Logistic regression trained\n")

# Q16: Calculate accuracy
cat("\n[Q16] Training accuracy\n")

pred_prob <- predict(model, model_data, type = "response")
pred <- ifelse(pred_prob > 0.5, 1, 0)
acc <- mean(pred == train$Survived)
cat("Accuracy:", sprintf("%.4f", acc), "\n")

# Q17: Process test data
cat("\n[Q17] Test data processing\n")

if (!file.exists("src/data/test.csv")) {
    cat("Error: test.csv not found\n")
    quit(status = 1)
}

test <- read_csv("src/data/test.csv", show_col_types = FALSE)
cat("Test shape:", paste(dim(test), collapse = " x "), "\n")

test$FamilySize <- test$SibSp + test$Parch + 1
test$IsAlone <- as.numeric(test$FamilySize == 1)
test$Sex_num <- as.numeric(test$Sex == "male")

test$Age[is.na(test$Age)] <- median(train$Age, na.rm = TRUE)
test$Fare[is.na(test$Fare)] <- median(train$Fare, na.rm = TRUE)
test$Embarked[is.na(test$Embarked)] <- names(sort(table(train$Embarked), decreasing = TRUE))[1]

cat("Applied transformations\n")

# Q18: Generate predictions
cat("\n[Q18] Predictions generated\n")

test_features <- test[features]
pred_prob <- predict(model, test_features, type = "response")
predictions <- ifelse(pred_prob > 0.5, 1, 0)

submission <- data.frame(
    PassengerId = test$PassengerId,
    Survived = predictions
)

if (!dir.exists("outputs")) {
    dir.create("outputs")
}

write.csv(submission, "outputs/r_predictions.csv", row.names = FALSE)

cat("Predictions for", length(predictions), "passengers\n")
cat("Survival rate:", sprintf("%.3f", mean(predictions)), "\n")
cat("Saved to outputs/r_predictions.csv\n")

cat("\nR analysis completed\n")
